elements/lisk-codec/test/decode.spec.ts

```diff
@@ -129,6 +149,180 @@ describe('decode', () => {
+			const schema = {
+				$id: 'duplicate/nonSequential',
+				type: 'object',
+				properties: {
+					later: {
+						dataType: 'uint32',
+						fieldNumber: 3,
+					},
+					data: {
+						dataType: 'uint32',
+						fieldNumber: 1,
+					},
+				},
+			};
+			const value = { later: 99, data: 255 };
+			const encoded = codec.encode(schema, value);
+
+			expect(codec.decode(schema, encoded)).toEqual(value);
+		});
+
+		it('should not decode a binary message where one required property (not of type array) is not contained in the binary message ', () => {
+			const schema = {
+				$id: 'duplicate/skipped',
+				type: 'object',
+				properties: {
+					data: {
+						dataType: 'uint32',
+						fieldNumber: 1,
+					},
+					later: {
+						dataType: 'uint32',
+						fieldNumber: 3,
+					},
+				},
+			};
+			const key = generateKey(schema.properties.data);
+			const intVal1 = writeUInt32(20);
+
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const encodedValue = Buffer.concat([keypair1]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow(
+				'Message does not contain a property for fieldNumber: 3.',
+			);
+		});
+
+		it('should not decode a binary message that contains a field number not present in the schema', () => {
+			const schema = {
+				$id: 'duplicate/fieldNumberNotExist',
+				type: 'object',
+				properties: {
+					data: {
+						dataType: 'uint32',
+						fieldNumber: 1,
+					},
+				},
+			};
+			const key = generateKey(schema.properties.data);
+			const intVal1 = writeUInt32(20);
+
+			const key2 = generateKey({
+				fieldNumber: 99,
+				dataType: 'uint64',
+			});
+			const intVal2 = writeUInt64(BigInt(9999));
+
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const keypair2 = Buffer.concat([key2, intVal2]);
+			const encodedValue = Buffer.concat([keypair2, keypair1]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow(
+				'Invalid field number while decoding',
+			);
+		});
+
+		it('should not decode a binary message that contains some bytes at the end that cannot be decoded to some key-value pairs. E.g., a binary message contatenated with the byte 0x00', () => {
+			const schema = {
+				$id: 'duplicate/extraBytes',
+				type: 'object',
+				properties: {
+					data: {
+						dataType: 'uint32',
+						fieldNumber: 1,
+					},
+				},
+			};
+			const key = generateKey(schema.properties.data);
+			const intVal1 = writeUInt32(20);
+
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const encodedValue = Buffer.concat([keypair1, Buffer.alloc(20, 255)]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow('Invalid terminate index.');
+		});
+
+		it('should not decode a binary message which contains a key-value pair for a property of type array for which packed encoding is used (for example boolean), and the value exist as empty bytes', () => {
+			const schema = {
+				$id: 'duplicate/emptyBytes',
```

Comment:
This schema also includes non-sequential fieldNumber, maybe to only have empty bytes case